File: app/controllers/EventSourceController.java
/*
 * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
 */
package controllers;

import akka.stream.javadsl.Source;
import com.google.common.collect.Lists;
import models.repositories.FluxRepository;
import models.repositories.ScreenRepository;
import play.libs.EventSource;
import play.mvc.Controller;
import play.mvc.Http;
import play.mvc.Result;
import services.FluxManager;
import views.html.eventsource;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.time.Duration;
import java.util.*;

@Singleton
public class EventSourceController extends Controller implements Observer {

    @Inject
    FluxRepository fluxRepository;

    @Inject
    ScreenRepository screenRepository;

    private static Source<String, ?> source;
    private Source<String, ?> oldSource;

    private boolean updated = false;

    @Inject
    EventSourceController() {

        FluxManager fluxManager = FluxManager.getInstance();
        fluxManager.addObserver(this);

        // TODO maybe optimize ?
        Thread t = new Thread(fluxManager);
        // t.start();
    }

    public Result index() {
        return ok(eventsource.render())
            .withHeader("Access-Control-Allow-Credentials", "true")
            .withHeader("Access-Control-Expose-Headers", "*")
            .withHeader("Access-Control-Allow-Origin", "localhost");
    }

    private static Http.Cookie[] toArray(final Http.Cookies cookies) {
    	Http.Cookie[] cookieArray = new Http.Cookie[1];
		return Lists.newArrayList(cookies.iterator()).toArray(cookieArray);
	}
    @Override
    @SuppressWarnings("unchecked")
    public synchronized void update(Observable o, Object arg) {

        if (source == null) {
            source = Source.tick(Duration.ZERO, Duration.ofSeconds(5), "tick");
        }
        else {
            List<String> list = new ArrayList<>();
            list.add((String) arg);
            Source<String, ?> s = Source.from(list);
            // source.merge(s);
            // source.flatMapMerge(1, s);
        }
    	//source = Source.single((String) arg);
        // System.out.println("Source was updated");
    	updated = true;
    }

    public Result events() {

        // TODO error page
        while (source != null) {
            // System.out.println("source is null");
        }

        //final Source<EventSource.Event, ?> eventSource;

        return ok().chunked(source
            //.map(t -> EventSource.Event.event(t.concat("\nretry: 10000\\n")))
            .map(EventSource.Event::event)
            .via(EventSource.flow()))
            .as(Http.MimeTypes.EVENT_STREAM);

        /*
        final Result bkp = ok();

        Result result = ok().chunked(eventSource.via(EventSource.flow()));
        if (bkp.flash() != null)
        	result = result.withFlash(bkp.flash());
        if (bkp.cookies() != null) {
			Iterator<Http.Cookie> iterator = bkp.cookies().iterator();
			while (iterator.hasNext())
				result = result.withCookies(iterator.next());
		}
        if (bkp.session() != null)
        	result = result.withSession(bkp.session());
        return result.as(Http.MimeTypes.EVENT_STREAM);*/

    }
}

File: app/controllers/EventSourceController.scala
package controllers

import akka.actor.{Actor, ActorRef, ActorSystem, Props}
import akka.stream.OverflowStrategy
import akka.stream.scaladsl.Source
import controllers.EventActorManager.{Register, SendMessage, UnRegister}
import javax.inject.{Inject, Singleton}
import play.api.http.ContentTypes
import play.api.libs.EventSource.EventNameExtractor
import play.api.mvc._
import play.libs.EventSource
import play.mvc.Result
import views.html.eventsource

import scala.collection.mutable
import scala.concurrent.ExecutionContext

/**
  * This class implements a controller for the Eventsources.
  * It uses Akka Actors to represent the clients (screens) and
  * register them to its "send-list"
  */
@Singleton
class EventSourceController @Inject() (system: ActorSystem,
                                        cc: ControllerComponents)
                                       (implicit executionContext: ExecutionContext)
extends AbstractController(cc) {


  private[this] val manager = system.actorOf(EventActorManager.props)

  implicit def pair[E]: EventNameExtractor[E] = EventNameExtractor[E](_ => Some("test1"))

  def send(event: String) =  {
    print("Send event to screens : " + event)
    manager ! SendMessage(event)
    Ok
  }

  def index = Action {
    Ok(eventsource.render())
  }
  def events = Action {

    val source  =
      Source
        .actorRef[String](32, OverflowStrategy.dropHead)
        .watchTermination() { case (actorRef, terminate) =>
          manager ! Register(actorRef)
          terminate.onComplete(_ => manager ! UnRegister(actorRef))
          actorRef
        }

    val eventSource = Source.fromGraph(source.map(EventSource.Event.event))

    Ok.chunked(eventSource via EventSource.flow).as(ContentTypes.EVENT_STREAM)
  }
}

class EventActor extends Actor {

  private[this] val actors = mutable.Set.empty[ActorRef]

  def receive = {
    case Register(actorRef)   => actors += actorRef
    case UnRegister(actorRef) => actors -= actorRef
    case SendMessage(message) => actors.foreach(_ ! message)
  }
}

object EventActorManager {
  def props: Props = Props[EventActor]

  case class SendMessage(message: String)

  case class Register(actorRef: ActorRef)
  case class UnRegister(actorRef: ActorRef)
}


